### PR TITLE
MTF dataset and packing

### PR DESCRIPTION
**Pull request by thomasw21**

### CODE REVIEW

File: megatron/data/data_samplers.py [hunk 1]
```diff
@@ -15,11 +15,77 @@
...
-from megatron import get_args
+
+from megatron import get_args, get_tokenizer
 from megatron import mpu
+from megatron.data.mtf_dataset import MTFDataset
+
+
+def pack_samples(items, max_seq_len: int, micro_batch_size: int, pad_token: int):
+    """
+    Greedily packs samples.
```
Comment (thomasw21) [inline megatron/data/data_samplers.py:30]: We should double check that their packing strategy is indeed greedy.

File: megatron/data/data_samplers.py [hunk 2]
```diff
@@ -15,11 +15,77 @@
...
+            }
+        ]
+    
+    Output:
+        decoder_target_tokens = [[6, 7, 8, 3, 4, 5, <pad>]]: Concatenation of tokens followed with padding tokens.
+        decoder_segment_ids = [[1, 1, 1, 2, 2, 2, 0]]: Segment ids determine original documents.
+        decoder_causal_attention = [[1, 1, 0, 1, 1, 0, 0]]: `0` depicts inputs, `1` depicts target.
+    """
+
+    decoder_target_tokens = np.full((micro_batch_size, max_seq_len), pad_token)
```
Comment (Muennighoff) [inline megatron/data/data_samplers.py:50]: We'll need to change `get_packed_attention_mask` 
from
`padding_mask = (tokens > 0).unsqueeze(-1) * (tokens > 0).unsqueeze(1)`
to 
`padding_mask = (tokens != pad_token).unsqueeze(-1) * (tokens != pad_token).unsqueeze(1)`

File: tests/test_dataloaders.py [hunk 3]
```diff
@@ -106,3 +108,91 @@ def test_mlm_dataset(self):
...
+                    # `seq_length` correspond to the one in argument + 1 in order to get tokens/labels
+                    self.assertEqual(seq_length, args.seq_length + 1)
+
+                    original_samples_count = 0
+                    for batch_id in range(micro_batch_size):
+                        segment_ids = [k for k, _ in itertools.groupby(items["decoder_segment_ids"][batch_id])]
+                        # `segment_ids` is [1,2,...]
+                        self.assertEqual(segment_ids[:-1], list(range(1, len(segment_ids))))
+                        # `0` signify that the tokens are padding
+                        self.assertEqual(segment_ids[-1], 0)
```
Comment (Muennighoff) [inline tests/test_dataloaders.py:185]: Why does the last segment id have to be padding?

File: tests/test_dataloaders.py [hunk 4]
```diff
@@ -106,3 +108,91 @@ def test_mlm_dataset(self):
...
+                        original_samples_count += len([segment_id for segment_id in segment_ids if segment_id != 0])
```
Comment (Muennighoff) [inline tests/test_dataloaders.py:186]:
```suggestion
                        original_samples_count += max(segment_ids)
```

File: megatron/data/data_samplers.py (hunk 1 above)
Comment (Muennighoff) [reply megatron/data/data_samplers.py:30]: The other option being some form of beam search to make padding as small as possible?

Comment (thomasw21) [reply megatron/data/data_samplers.py:30]: Discussed offline, we should check the other implementations, I don't have a good idea of who they do it, but I expect them to have something better than greedy packing.

File: tests/test_dataloaders.py (hunk 3 above)
Comment (thomasw21) [reply tests/test_dataloaders.py:185]: You're right, I'll fix them.

Comment (thomasw21) [reply tests/test_dataloaders.py:185]: Done in [c74dbb7](https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/293/commits/c74dbb7ba6ae593ec6c391fc8f3e08fa7651e791)

File: megatron/data/data_samplers.py [hunk 5]
```diff
@@ -15,11 +15,77 @@
...
+
+        item_num += 1
+        cur_len += total_len
+        assert cur_len < max_seq_len
+
+    return {
+        "decoder_target_tokens": decoder_target_tokens,
+        "decoder_segment_ids": decoder_segment_ids,
+        "decoder_causal_attention": decoder_causal_attention,
+    }
```
Comment (Muennighoff) [inline megatron/data/data_samplers.py:88]: Still need to cast them to torch tensors / directly use torch tensors instead of numpy

File: tests/test_dataloaders.py [hunk 6]
```diff
@@ -106,3 +108,96 @@ def test_mlm_dataset(self):
...
+                    data_prefix=args.data_path,
+                    data_impl=args.data_impl,
+                    splits_string=args.split,
+                    # TODO @thomasw21 figure how that value works
+                    train_valid_test_num_samples=train_val_test_num_samples,
+                    seed=args.seed,
+                    skip_warmup=(not args.mmap_warmup)
+                )
+
+                # TODO @thomasw21 make sure that input and target are aligned.
```
Comment (Muennighoff) [inline tests/test_dataloaders.py:137]: Note: We will check this when preprocessing

**Review by Muennighoff — APPROVED** [review]

File: megatron/data/data_samplers.py [hunk 7]
```diff
@@ -158,3 +245,76 @@ def __iter__(self):
...
+        random_idx = torch.randperm(bucket_size, generator=g).tolist()
+        idx_range = [start_idx + x for x in random_idx[bucket_offset:]]
+
+        batch = []
+        batch_count = 0
+        token_lens = 0
+        # Last batch if not complete will be dropped.
+        for idx in idx_range:
+            tok_len = len(self.dataset[idx]['input_tokens']) + len(self.dataset[idx]['target_tokens'])
+            if token_lens + tok_len > self.sequence_length:
```
Comment (thomasw21) [inline megatron/data/data_samplers.py:309]: I think a good opportunity to have the EOS token that we didn't have from pretrianing is adding one here.
```
{INPUT} {TARGET} <EOS>
```
Not sure we should have a <EOS> between input and target
WDYT @TevenLeScao @Muennighoff @lintangsutawika

File: megatron/data/data_samplers.py [hunk 8]
```diff
@@ -158,3 +245,76 @@ def __iter__(self):
...
+        # data sharding and random sampling
+        bucket_size = (self.total_samples // self.micro_batch_times_data_parallel_size) \
+                      * self.micro_batch_size
+        bucket_offset = current_epoch_samples // self.data_parallel_size
+        start_idx = self.data_parallel_rank * bucket_size
+
+        g = torch.Generator()
+        g.manual_seed(self.epoch)
+
+        random_idx = torch.randperm(bucket_size, generator=g).tolist()
```
Comment (thomasw21) [inline megatron/data/data_samplers.py:300]: I haven't thought about it very clearly, but essentially this mean that the entire dataset can be random accessed which isn't possible if you pack things greedily (the problem is given an index, can you know which original sample you need to add in the batches ....) So I'm refactoring this to put this in another dataset ... Sorry about this @Muennighoff you were right.

File: megatron/data/data_samplers.py (hunk 7 above)
Comment (Muennighoff) [reply megatron/data/data_samplers.py:309]: Good idea!
I think we should add an EOS token only after each target

Comment (thomasw21) [reply megatron/data/data_samplers.py:309]: Yeah agree on only after target.

Comment (TevenLeScao) [reply megatron/data/data_samplers.py:309]: Agreed

**Comment by stas00** [comment]
FYI, `tests/test_dataloaders.py::TestDataLoading::test_mlm_dataset` fails. 

I've added a skip to not run it by CI please fix at your convenience.

I partially skipped it - it was never really working as the args were invalid.

If fails on CI and on my desktop as well:

```
>               train_ds, valid_ds, test_ds = mlm_dataset.build_train_valid_test_datasets(
                    data_prefix=args.data_path,
                    data_impl=args.data_impl,
                    splits_string=args.split,
                    # TODO @thomasw21 figure how that value works
                    train_valid_test_num_samples=train_val_test_num_samples,
                    sequence_length=args.seq_length,
                    noise_density=args.noise_density,
                    mean_noise_span_length=args.mean_noise_span_length,
                    seed=args.seed,
                    skip_warmup=(not args.mmap_warmup)
                )

tests/test_dataloaders.py:90: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
megatron/data/mlm_dataset.py:25: in build_train_valid_test_datasets
    return _build_train_valid_test_datasets(
megatron/data/mlm_dataset.py:263: in _build_train_valid_test_datasets
    train_dataset = build_dataset(0, 'train')
megatron/data/mlm_dataset.py:250: in build_dataset
    dataset = MLMDataset(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <megatron.data.mlm_dataset.MLMDataset object at 0x7fc94710c070>, name = 'train'
indexed_dataset = <megatron.data.indexed_dataset.MMapIndexedDataset object at 0x7fc94710c880>
documents = array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,...9, 930, 931, 932, 933, 934, 935,
       936, 937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947],
      dtype=int32)
data_prefix = '/mnt/nvme0/code/huggingface/Megatron-DeepSpeed-master/tests/data/gpt2/meg-gpt2-openwebtext_text_document'
sequence_length = 512, num_samples = 40000, seed = 1234, noise_density = 0.15, mean_noise_span_length = 3

    def __init__(
        self,
        name,
        indexed_dataset,
        documents,
        data_prefix,
        sequence_length,
        num_samples,
        seed,
        noise_density=0.15,
        mean_noise_span_length=3
    ):
    
        # Params to store.
        self.name = name
        self.seed = seed
        self.sequence_length = sequence_length
    
        # Dataset.
        self.indexed_dataset = indexed_dataset
    
        self.noise_density = noise_density
        self.mean_noise_span_length = mean_noise_span_length
        # T5-like span masked language modeling will fuse consecutively masked tokens to a single sentinel token.
        # To ensure that the input length is `sequence_length`, we need to increase the maximum length
        # according to `noise_density` and `mean_noise_span_length`. We can also define the label length accordingly.
        number_of_raw_tokens, inputs_length, targets_length, num_noise_spans = compute_input_and_target_lengths(
            sequence_length=self.sequence_length,
            noise_density=self.noise_density,
            mean_noise_span_length=self.mean_noise_span_length
        )
        self.inputs_length = inputs_length
        # In order to compute loss, we need an extra token at the end.
        self.number_of_raw_tokens = number_of_raw_tokens + 1
        self.targets_length = targets_length + 1
        self.num_noise_spans = num_noise_spans
    
        # Build the samples mapping.
        self._gpt_dataset = GPTDataset(
            name=self.name,
            data_prefix=data_prefix,
            documents=documents,
            indexed_dataset=self.indexed_dataset,
            num_samples=num_samples,
            # -1 because GPTDataset will return `seq_length + 1` sequences.
            seq_length=number_of_raw_tokens - 1,
            seed=seed
        )
    
        # Vocab stuff.
        tokenizer = get_tokenizer()
        self.sep_id = tokenizer.sep
        self.sentinel_token_ids = tokenizer.additional_special_tokens_ids
        assert self.sep_id is not None, "MLM dataset requires tokenizer to have a <sep> token"
        assert len(self.sentinel_token_ids) > 0, "Provide the argument --vocab-extra-ids 100 to the script"
        assert len(self.sentinel_token_ids) >= self.num_noise_spans, "Not enough sentinel tokens, please add more"
    
        args = get_args()
        if hasattr(args, "encoder_seq_length") and args.encoder_seq_length is not None:
            # T5 style
>           assert self.inputs_length == args.encoder_seq_length
E           AssertionError

megatron/data/mlm_dataset.py:332: AssertionError
```
